blog/2020-02-28-db2-cdc-approaches.adoc

```diff
@@ -0,0 +1,248 @@
+the remote database an apply agent periodically reads the changes from
+these CDC tables and updates the shadow tables.
+
+While conceptually this is quite simple in practice it is difficult to
+automate  for the  following reasons: 
+
+ - Source  and sink  are tightly coupled  and therefore the same  table cannot  easily be replicated  to multiple  different target  database system.  
+-  If  the source  system  was already  using replication on a  table e.g. for back-up purposes, we  cannot use this method to replicate to the Datalake. 
+- Elevated privileges are required on the source. Data owners give read access to their system for Sqoop, but  giving administrator  access  to  a third  party  system  is a  compliance nightmare.   
+-  Elevated privileges  are  required  on the  sink.  For simplicity  our  Drop Zone  is  a  single  Db2 system  with  database instances for each  of the data sources.  Allowing the  data owners to set up  SQL Replication to  the Drop Zone  would allow them  access to each other  systems, which is  a compliance violation and  a management nightmare.  
+- The  tools are designed for system admins and as a result  there are a large number  of gotchas  for  the  unwary. For example, care must be taken in selecting a wide range of parameters such as: the mode  that  the transaction log has  to be in to  allow CDC, the time  the last backup was taken, whether the  database is row or column oriented etc.  
+- It is a Db2-specific  solution, although the  majority of the  relational data
+sources were Db2, we also had Netezza, MySQL and SQL Server sources.
+
+
+We found in  practice that the combination of the  above meant that it
+was impractical to  allow data owners to use IBM  SQL Replication as a
+CDC mechanism for the Datalake.
+
+IBM  offers another  set  of  tools for  data  replication called  IBM
+InfoSphere Data Replication (link:https://www.ibm.com/support/knowledgecenter/SSTRGZ_11.4.0/com.ibm.idr.frontend.doc/pv_welcome.html[IIDR]). This is sold as a product distinct
+from Db2.   IIDR is not  a Db2 specific  solution, working for  a wide
+range of relational  databases as well as  non-relational data storage
+systems, e.g. file systems.  In essence IIDR has source agents and sink
+agents.   The source  and sink  agent run  at or  close to  the target
+system. Source agents  read the changes and propagate them  via a wide
+range of protocols  including TCP sockets, MQ, Shared Files  etc. to the
+sink agent.  The  source and sink agents are configured  via an entity
+called the Access Server through which sources are  connected to sinks
+and the tables  to capture are specified. The Access Server is itself
+typically  controlled via  a  Graphical User  Interface  by a  system
+administrator.
+
+Thus for example  we can have a  Db2 source agent and  a Kafka Confluent
+sink agent.  Through configuration of the Access Server a table can be
+captured at the source and a Kafka Table (or KTable) representation of
```

Comment:
It is a Kafka Confluent Streams KTable

Reply:
Sorry, I still don't quite get it :) What is "Kafka Confluent Streams"? I think we should focus as much as possible on the Apache Kafka upstream project, in particular there's no such thing like "Confluent Kafka".

Reply:
Confluent (the company)  releases  a product called Kafka Streams. As part of that they introduce a representation of a relational table in Kafka called a KTable. Please tell me the least ambiguous formulation of that concept.

Reply:
Ah, ok, I begin to see where the confusion is coming from then.

Kafka Streams is *not* a product of Confluent, but [part of](https://kafka.apache.org/24/documentation/streams/) the Apache Kafka upstream project. A `KTable` is an element in the Kafka Streams API, which for all the records in a Kafka topic with the same key provides a view on the latest record with that key ("upsert" semantics).

That said, I don't quite understand how Kafka Streams and its `KTable` concept are leveraged in the IIDR architecture; is the intention to rate limit the emit frequency of change events exposed to downstream consumers? In other words, why don't consumers directly receive the change events produced by IIDR? That's what we typically see in the Debezium community, with Kafka Streams optionally being used for _stream processing_, like transforming or enriching change events.

How about instead of this:

> ~~Thus for example  we can have a  Db2 source agent and an IIDR Kafka sink agent
that behaves like a standard Kafka Confluent source connector, i.e.
it updates a Confluent Schema Registry and writes into a Confluent Kafka
Table (KTable). The initial rows  of the table
being recorded  as a set  of Upsert messages (REFRESH phase)  and subsequent
changes to  the table  being recorded as  a sequence  of Upsert/Delete
Messages (MIRROR phase).~~
>
> ~~The KTable can  itself be streamed into a target  database in the Drop
Zone or used to make a snapshot that is loaded directly into the Landing
Zone.~~

the following, simply leaving all those details aside:

> Thus for example  we can have a Db2 source agent and an IIDR Kafka sink agent
that behaves like a standard Apache Kafka Connect source connector, i.e. it writes change events into a Kafka topic. The initial records are Upsert messages (REFRESH phase)  and subsequent changes are propagated as a sequence  of Upsert/Delete
Messages (MIRROR phase).

If you actually want to touch on the notion of `KTable`, its usage in the overall architecture should be motivated.

Reply:
No particular desire to talk about KTables so fine with me.

Reply:
Ok, cool. Can you change it into this then? Otherwise I think it's good to go. Can we publish it tomorrow, or do you still need approval from your org?